src/native/http2_stream_manager.c

```diff
@@ -0,0 +1,403 @@
+        aws_jni_throw_runtime_exception(env, "Failed to create stream manager: %s", aws_error_str(aws_last_error()));
+    }
+
+    aws_http_proxy_options_jni_clean_up(
+        env, &proxy_options, jni_proxy_host, jni_proxy_authorization_username, jni_proxy_authorization_password);
+
+    if (use_tls) {
+        aws_tls_connection_options_clean_up(&tls_conn_options);
+    }
+
+cleanup:
+    aws_jni_byte_cursor_from_jbyteArray_release(env, jni_endpoint, endpoint);
+
+    if (binding->stream_manager == NULL) {
+        s_destroy_manager_binding(binding);
+        binding = NULL;
+    }
+
+    return (jlong)binding;
+}
+
+/*
+ * Stream manager binding, persists across the lifetime of the native object.
+ */
+struct aws_sm_acquire_stream_callback_data {
+    JavaVM *jvm;
+    struct http_stream_binding *stream_binding;
+    jobject java_async_callback;
+};
+
+static void s_cleanup_sm_acquire_stream_callback_data(struct aws_sm_acquire_stream_callback_data *callback_data) {
+
+    JNIEnv *env = aws_jni_get_thread_env(callback_data->jvm);
+
+    if (callback_data->java_async_callback) {
+        (*env)->DeleteGlobalRef(env, callback_data->java_async_callback);
+    }
+    aws_mem_release(aws_jni_get_allocator(), callback_data);
+}
+
+static struct aws_sm_acquire_stream_callback_data *s_new_sm_acquire_stream_callback_data(
+    JNIEnv *env,
+    struct aws_allocator *allocator,
+    struct http_stream_binding *stream_binding,
+    jobject async_callback) {
+    struct aws_sm_acquire_stream_callback_data *callback_data =
+        aws_mem_calloc(allocator, 1, sizeof(struct aws_sm_acquire_stream_callback_data));
+
+    jint jvmresult = (*env)->GetJavaVM(env, &callback_data->jvm);
+    AWS_FATAL_ASSERT(jvmresult == 0);
+    callback_data->java_async_callback = async_callback ? (*env)->NewGlobalRef(env, async_callback) : NULL;
+    AWS_FATAL_ASSERT(callback_data->java_async_callback != NULL);
+    callback_data->stream_binding = stream_binding;
+
+    return callback_data;
+}
+
+static void s_on_stream_acquired(struct aws_http_stream *stream, int error_code, void *user_data) {
+    struct aws_sm_acquire_stream_callback_data *callback_data = user_data;
+    JNIEnv *env = aws_jni_get_thread_env(callback_data->jvm);
+    if (error_code) {
+        jobject crt_exception = aws_jni_new_crt_exception_from_error_code(env, error_code);
+        (*env)->CallVoidMethod(
+            env, callback_data->java_async_callback, async_callback_properties.on_failure, crt_exception);
+        (*env)->DeleteLocalRef(env, crt_exception);
+        aws_http_stream_binding_destroy(env, callback_data->stream_binding);
+    } else {
+        callback_data->stream_binding->native_stream = stream;
+        jobject j_http_stream =
+            aws_java_http_stream_from_native_new(env, callback_data->stream_binding, AWS_HTTP_VERSION_2);
+        if (!j_http_stream) {
+            jobject crt_exception = aws_jni_new_crt_exception_from_error_code(env, aws_last_error());
```

Comment:
if `aws_java_http_stream_from_native_new()` fails it doesn't set an aws-error-code
it raises a java exception

Reply:
updated to get the exception using `ExceptionOccurred`